### PR TITLE
feat: support for having port in host header when pass_host = node and port is not standard

### LINKED ISSUE #6366: feat: As a User, I want to have the port in the Host header when `pass_host = node` and the port is not standard
### Issue description

feat: As a User, I want to have the port in the Host header when `pass_host = node` and the port is not standard, so that we can solve the problem in https://github.com/apache/apisix/issues/6296 in a better way when:

1. the port in the Host header is checked
2. multiple nodes which have different host name

### PR DESCRIPTION
**Pull request by wilson-1024**
### What this PR does / why we need it:

support for having port in host header when pass_host = node and port is not standard

resolve #6366

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: t/cli/test_standalone.sh [hunk 1]
```diff
@@ -65,3 +65,139 @@ if [ ! $code -eq 200 ]; then
 fi
 
 echo "passed: resolve variables in apisix.yaml conf success"
+
```
Comment (spacewander) [inline t/cli/test_standalone.sh:68]: This file is not suitable to add this test.
Please add it in node/upstream.t like https://github.com/apache/apisix/blob/a490e0a2570d916921b4a4e50f3c308291ba7dd9/t/node/upstream.t#L400

Comment (wilson-1024) [reply t/cli/test_standalone.sh]: Fixed. Please review it again.

File: t/lib/server.lua [hunk 2]
```diff
@@ -531,6 +531,11 @@ function _M.plugin_proxy_rewrite_resp_header()
     ngx.say(s)
 end
 
+function _M.server_upstream_host()
```
Comment (spacewander) [inline t/lib/server.lua:534]: There is no need to add a new backend. `/uri` already contains the host: https://github.com/apache/apisix/blob/530fb4d62c990ec23026f43f70b878e157d441bc/t/node/upstream.t#L510

File: t/node/upstream.t [hunk 3]
```diff
@@ -510,3 +510,54 @@ GET /uri
...
+                        },
+                        "upstream": {
+                            "nodes": {
+                                "127.0.0.1:1979": 1,
+                                "127.0.0.1:1980": 1000
+                            },
+                            "type": "roundrobin",
+                            "pass_host": "node"
+                        },
+                        "uri": "/hello"
```
Comment (spacewander) [inline t/node/upstream.t:538]: We can use `"/uri"` directly, as the test I pointed out before

File: t/node/upstream.t [hunk 4]
```diff
@@ -510,3 +510,54 @@ GET /uri
...
+
+
+
+=== TEST 22: hit route
+--- request
+GET /hello
+--- response_body
+127.0.0.1:1980
+--- no_error_log
+[error]
```
Comment (spacewander) [inline t/node/upstream.t:563]: Also need to cover the case with retry: https://github.com/apache/apisix/blob/530fb4d62c990ec23026f43f70b878e157d441bc/t/node/upstream.t#L458

```
"nodes": {
                        "127.0.0.1:1979": 1000,
                        "localhost:1980": 1
                    }
```
We use 1979, a nonexistent port to trigger the retry.

Comment (wilson-1024) [reply t/node/upstream.t:618]: Thank you.I have solved the problem mentioned,please review it again.

File: t/node/upstream.t [hunk 5]
```diff
@@ -510,3 +510,125 @@ GET /uri
...
+
+
+
+=== TEST 22: hit route
+--- request
+GET /uri
+--- skip_nginx: 5: < 1.19.0
+--- response_body
```
Comment (spacewander) [inline t/node/upstream.t:568]: we can use:
```
--- response_body eval
qr/host: localhost:1980/
```

**Comment by spacewander** [comment]
@wilson-1024 
You can apply:
```
diff --git t/node/upstream.t t/node/upstream.t
index feee95bd..12b7123d 100644
--- t/node/upstream.t
+++ t/node/upstream.t
@@ -565,10 +565,8 @@ passed
 --- request
 GET /uri
 --- skip_nginx: 5: < 1.19.0
---- response_body
-uri: /uri
-host: localhost:1980
-x-real-ip: 127.0.0.1
+--- response_body eval
+qr/host: localhost:1980/
 --- no_error_log
 [error]

@@ -623,12 +621,9 @@ passed


 === TEST 24: hit route
+--- log_level: debug
 --- request
 GET /uri
 --- skip_nginx: 5: < 1.19.0
---- response_body
-uri: /uri
-host: localhost:1980
-x-real-ip: 127.0.0.1
 --- error_log
-proxy request to 127.0.0.1:1980
+Host: 127.0.0.1:1979
```

**Comment by spacewander** [comment]
@wilson-1024 
There are other issues in the test file. Would be better to apply my patch directly.

**Comment by wilson-1024** [comment]
@spacewander Sorry, I didn't read your comment carefully just now.I am modifying to apply your patch directly.

**Comment by wilson-1024** [comment]
@spacewander  I applied this patch directly, but my local test fails.Please help me.
result：
```
 FLUSH_ETCD=1 prove -I. -I../test-nginx/ -I../test-nginx/lib -r  t/node/upstream.t
t/node/upstream.t .. 66/? 
#   Failed test 't/node/upstream.t TEST 24: hit route - pattern "Host: 127.0.0.1:1979" should match a line in error.log (req 0)'
#   at /Users/zenghuilin/github/apisix/../test-nginx/lib/Test/Nginx/Socket.pm line 1284.
# Looks like you failed 1 test of 71.
t/node/upstream.t .. Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/71 subtests 

Test Summary Report
-------------------
t/node/upstream.t (Wstat: 256 Tests: 71 Failed: 1)
  Failed test:  71
  Non-zero exit status: 1
Files=1, Tests=71, 12 wallclock secs ( 0.03 usr  0.00 sys +  1.52 cusr  0.89 csys =  2.44 CPU)
Result: FAIL
```
file diff
```diff
diff --git a/t/node/upstream.t b/t/node/upstream.t
index feee95bd..c73b8391 100644
--- a/t/node/upstream.t
+++ b/t/node/upstream.t
@@ -565,10 +565,8 @@ passed
 --- request
 GET /uri
 --- skip_nginx: 5: < 1.19.0
---- response_body
-uri: /uri
-host: localhost:1980
-x-real-ip: 127.0.0.1
+--- response_body eval
+qr/host: localhost:1980/
 --- no_error_log
 [error]
 
@@ -614,7 +612,6 @@ x-real-ip: 127.0.0.1
     }
 --- request
 GET /t
---- skip_nginx: 5: < 1.19.0
 --- response_body
 passed
 --- no_error_log
@@ -623,12 +620,10 @@ passed
 
 
 === TEST 24: hit route
+--- log_level: debug
 --- request
 GET /uri
---- skip_nginx: 5: < 1.19.0
---- response_body
-uri: /uri
-host: localhost:1980
-x-real-ip: 127.0.0.1
+--- response_body eval
+qr/host: localhost:1980/
 --- error_log
-proxy request to 127.0.0.1:1980
+Host: 127.0.0.1:1979

```

**Comment by spacewander** [comment]
Is there `--with-debug` in your `nginx -V`? Debug log is not enabled by default.

**Comment by wilson-1024** [comment]
I think this can be used as a check for retry
apisix/apisix/balancer.lua  line 362 
```
core.log.info("proxy request to ", server.host, ":", server.port)
```

**Comment by wilson-1024** [comment]
I don't have --with-debug in your nginx -V

**Comment by wilson-1024** [comment]
I don't have --with-debug in my nginx -V

**Comment by spacewander** [comment]
> I think this can be used as a check for retry apisix/apisix/balancer.lua line 362
> 
> ```
> core.log.info("proxy request to ", server.host, ":", server.port)
> ```

The log here doesn't include the real Host header. That's why I use debug log.

Simply committing the patch is enough. The openresty in CI has debug log enabled.

**Comment by wilson-1024** [comment]
@spacewander 
Thank you for your careful instruction. When i enable the nginx debugging option,the local test succeeds.

```
FLUSH_ETCD=1 prove -I. -I../test-nginx/ -I../test-nginx/lib -r  t/node/upstream.t
t/node/upstream.t .. ok    
All tests successful.
Files=1, Tests=71, 17 wallclock secs ( 0.03 usr  0.01 sys +  2.16 cusr  1.22 csys =  3.42 CPU)
Result: PASS

```
log file
```
[debug] 7218#2263361: *23 http proxy header:
"GET /uri HTTP/1.1
Host: 127.0.0.1:1979
X-Real-IP: 127.0.0.1
X-Forwarded-For: 127.0.0.1
X-Forwarded-Proto: http
X-Forwarded-Host: localhost
X-Forwarded-Port: 1984
```

**Review by spacewander — COMMENTED** [review]
Sorry we should remove `--- skip_nginx: 5: < 1.19.0` in TEST 21~22, and keep it in TEST 23~24

**Comment by lijing-21** [comment]
Hi @wilson-1024  , thank you for your contribution!

Here is the Contributor T-shirt form[1], if you're interested, kindly take a look :)

[1] https://github.com/apache/apisix/blob/master/CONTRIBUTING.md#contributor-t-shirt

File: apisix/init.lua [hunk 6]
```diff
@@ -260,8 +260,8 @@ local function set_upstream_host(api_ctx, picked_server)
     if nodes_count == 1 then
         local node = up_conf.nodes[1]
         api_ctx.var.upstream_host = node.domain or node.host
```
Comment (spacewander) [inline apisix/init.lua:262]: We should also handle this path (nodes_count == 1)

Comment (wilson-1024) [reply apisix/init.lua:262]: Sorry, I have been busy with work and have no time to reply.ok, I'll deal with it later.

**Review by spacewander — COMMENTED** [review]
Err... Look like you need to update some tests as they expect "localhost" but not "localhost:port".
Feel free to do it when you have spare time.

**Comment by wilson-1024** [comment]
There is an error in the output of CI.
```
! Installing the dependencies failed: Module 'Test::Base' is not installed, Module 'Text::Diff' is not installed
! Bailing out the installation for Test-Nginx-0.29.
2 distributions installed
Error: Process completed with exit code 1.
```

**Review by spacewander — APPROVED** [review]

File: apisix/balancer.lua [hunk 7]
```diff
@@ -176,6 +177,17 @@ local function set_balancer_opts(route, ctx)
 end
 
 
+local function parse_server_for_upstream_host(picked_server, ctx)
+    local upstream_scheme = ctx.upstream_scheme
+    local standard_port = apisix_upstream.scheme_to_port[upstream_scheme]
```
Comment (tzssangglass) [inline apisix/balancer.lua:182]:
```suggestion
local function parse_server_for_upstream_host(picked_server, upstream_scheme)
    local standard_port = apisix_upstream.scheme_to_port[upstream_scheme]
```

is better?

Comment (wilson-1024) [reply apisix/balancer.lua]: Yes.Thank you for your careful instruction

**Review by spacewander — APPROVED** [review]

**Review by tzssangglass — APPROVED** [review]

**Pull request merged**